packages/tools/client-debugger/client-debugger-chrome-extension/src/react-components/TelemetryView.tsx

```diff
@@ -69,5 +68,23 @@ export function TelemetryView(): React.ReactElement {
 		};
 	}, [messageRelay, setTelemetryEvents]);
 
-	return <_TelemetryView telemetryEvents={telemetryEvents}></_TelemetryView>;
+	if (telemetryEvents.length > pageLimit) {
+		setTelemetryEvents(telemetryEvents.slice(0, pageLimit));
+	}
```

Comment:
Let's move this check to the places where we are already calling `setTelemetryEvents()`. Having it here outside the effect means that when new messages arrive, this will happen:
- The effect triggers.
- We update the component's state (`telemetryEvents`) with a list with 60 events (as an example).
- The component logic runs again for a re-render in response to the state change. As it is executing, we update the state again (with this call).
- The component renders _and displays 60 events_ (this call to `setTelemetryEvents()` doesn't change the value of `telemetryEvents` seen by the `return` below).
- The component immediately renders _again_ because of the state change outside the effect, now displaying the correct number.

We can avoid the last re-rendering by making sure that we always keep the max number of events when we update the list of events in state where we have to.

Reply:
I see. Now it move into the `_TelemetryView`. It should render when there is a new event comes in and call `setTelemetryEvents` and pass the latest `telemetryEvents` to `_TelemetryView` to render it on page. Should it only render once?

Reply:
Yeah, with the code as it is right now, we'll avoid the re-render 👍 .

Reply:
Awesome! Today I learned.